### PR TITLE
shaka-packager: init at 2.6.1

### PR DESCRIPTION
**Pull request by simonrw**
###### Description of changes

Add `shaka-packager` to `nixpkgs`, for media packaging.

https://shaka-project.github.io/shaka-packager/html/

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.05 Release Notes (or backporting 22.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by abathur** [comment]
It would be nice to have this.

A quick skim of the [build instructions](https://github.com/shaka-project/shaka-packager/blob/main/docs/source/build_instructions.md) persuaded me that repackaging the binaries is reasonable for now.

File: pkgs/applications/video/shaka-packager/default.nix
Comment (doronbehar) [inline pkgs/applications/video/shaka-packager/default.nix:42]: How about:

```suggestion
      ${shaka-packager}/bin/packager -version | grep ${version} > $out
```

Comment (simonrw) [reply pkgs/applications/video/shaka-packager/default.nix]: Thanks, that's a nice suggestion

**Comment by doronbehar** [comment]
Please test that the change suggested above works as expected, and also squash the commits.

File: pkgs/applications/video/shaka-packager/default.nix
Comment (AndersonTorres) [inline pkgs/applications/video/shaka-packager/default.nix:7]:
```suggestion
{ lib
, stdenv
, fetchurl
, runCommand
, shaka-packager
}:

```

File: pkgs/applications/video/shaka-packager/default.nix
Comment (AndersonTorres) [inline pkgs/applications/video/shaka-packager/default.nix:9]:
```suggestion
  sources = {
```

File: pkgs/applications/video/shaka-packager/default.nix
Comment (AndersonTorres) [inline pkgs/applications/video/shaka-packager/default.nix:13]:
```suggestion
    "x86_64-linux" = {
      filename = "packager-linux-x64";
      hash = "sha256-MoMX6PEtvPmloXJwRpnC2lHlT+tozsV4dmbCqweyyI0=";
    };
```

Run `nix hash to-sri sha256:13f8n83sphk6frwcbkk8xd7yalfsqaclcw3jl6jzkg1dy7l1g0rj` to modify to SRI hash.

File: pkgs/applications/video/shaka-packager/default.nix
Comment (AndersonTorres) [inline pkgs/applications/video/shaka-packager/default.nix:22]:
```suggestion
  }."${stdenv.hostPlatform.system}"
    or (throw "Unsupported system: ${stdenv.hostPlatform.system}");
```

File: pkgs/applications/video/shaka-packager/default.nix
Comment (AndersonTorres) [inline pkgs/applications/video/shaka-packager/default.nix:38]:
```suggestion
  installPhase = ''
    runHook preInstall
    
    install -m755 -D $src $out/bin/packager
    
    runHook postInstall
  '';
```

File: pkgs/applications/video/shaka-packager/default.nix
Comment (AndersonTorres) [inline pkgs/applications/video/shaka-packager/default.nix:52]:
```suggestion
  meta = {
    homepage = "https://shaka-project.github.io/shaka-packager/html/";
    description = "Media packaging framework for VOD and Live DASH and HLS applications";
    license = lib.licenses.bsd3;
    maintainers = with maintainers; [ ];
    platforms = [ "aarch64-linux" "x86_64-linux" "x86_64-darwin" ];
  };
```

File: pkgs/applications/video/shaka-packager/default.nix
Comment (AndersonTorres) [inline pkgs/applications/video/shaka-packager/default.nix:30]: Flip url and hash.

**Review by AndersonTorres — CHANGES_REQUESTED** [review]

**Review by doronbehar — CHANGES_REQUESTED** [review]

**Comment by doronbehar** [comment]
And please squash the commits every time you apply changes.

File: pkgs/applications/video/shaka-packager/default.nix
Comment (AndersonTorres) [inline pkgs/applications/video/shaka-packager/default.nix:16]: Convert all these hashes to SRI format.

**Review by AndersonTorres — CHANGES_REQUESTED** [review]

**Comment by simonrw** [comment]
Thanks @AndersonTorres and @doronbehar for your suggestions. I've made the requested changes

File: pkgs/applications/video/shaka-packager/default.nix
Comment (AndersonTorres) [inline pkgs/applications/video/shaka-packager/default.nix:40]: `packager` is too "generic".
I believe we should rename it to `shaka-packager`.
However, it can be problematic because I don't know how other distros name it.

(I always cite the hypothetical case of "someone that uses a script that expects the official naming".)

Edit: only aur, and the name is `packager` indeed.

File: pkgs/applications/video/shaka-packager/default.nix
Comment (AndersonTorres) [inline pkgs/applications/video/shaka-packager/default.nix:56]:
```suggestion
    mainProgram = "packager";
    maintainers = with lib.maintainers; [ ];
```

Sorry, I forgot this.

**Review by AndersonTorres — CHANGES_REQUESTED** [review]
Besides, it's ready to merge.

File: pkgs/applications/video/shaka-packager/default.nix
Comment (drupol) [inline pkgs/applications/video/shaka-packager/default.nix:25]: I think we should use the `finalAttrs` pattern here.

The `finalAttrs` pattern will let you remove the `rec` keyword ([see its implementation in Nix](https://github.com/NixOS/nix/blob/2d1d81114d72ace89ce08cd3bc93f4eb27a2975d/src/libexpr/nixexpr.cc#L121)).

Why is this pattern preferred to `rec` ?

Let's take this simple code example:

```nix
mkDerivation rec {
   foo = 1;
   bar = foo + 1;
}
```

and then `.overrideAttrs(old: { foo = 2; })`, you'll get `{ foo = 2; bar = 2; }` while with `finalAttrs` pattern, it would work correctly because it's a real fixed point.

Let me share a couple of useful links regarding the `finalAttrs` pattern:

1. History: https://discourse.nixos.org/t/avoid-rec-expresions-in-nixpkgs/8293
2. Documentation: https://nixos.org/manual/nixpkgs/unstable/#mkderivation-recursive-attributes
3. Recent example of implementation: https://github.com/NixOS/nixpkgs/compare/17f96f7b978e61576cfe16136eb418f74fab9952..9e6ea843e473d34d4f379b8b0d8ef0425a06defe

Feel free to reach out if you need some assistance.

File: pkgs/applications/video/shaka-packager/default.nix
Comment (drupol) [inline pkgs/applications/video/shaka-packager/default.nix:52]: While you're at it, could you sort the `meta` attributes alphabetically?

**Review by drupol — CHANGES_REQUESTED** [review]
Hi !

Thanks for your first contribution, welcome to the family !

I've left some comments, let me know if this is clear enough.
Feel free to request some assistance if something is too complex.

Thanks!

File: pkgs/applications/video/shaka-packager/default.nix
Comment (simonrw) [reply pkgs/applications/video/shaka-packager/default.nix]: Thanks, I only learnt about the `finalAttrs` pattern recently. I didn't realise it was recommended practice, but I'm glad it is!

File: pkgs/applications/video/shaka-packager/default.nix
Comment (simonrw) [reply pkgs/applications/video/shaka-packager/default.nix:40]: I'm not sure about this. I imagine when people install the package, they expect the tool to be called `packager` rather than anything else. Happy to trust your judgement and make the change if required though.

File: pkgs/applications/video/shaka-packager/default.nix
Comment (AndersonTorres) [reply pkgs/applications/video/shaka-packager/default.nix]: It is not recommended yet because this pattern was not ported outside `mkDerivation`.

**Review by AndersonTorres — APPROVED** [review]

File: pkgs/applications/video/shaka-packager/default.nix
Comment (abathur) [reply pkgs/applications/video/shaka-packager/default.nix:40]: I'm not sure if it's the ~best way to check for existing clashes are, but FWIW `nix run github:mic92/nix-index-database /bin/packager` doesn't indicate any.

File: pkgs/applications/video/shaka-packager/default.nix
Comment (doronbehar) [inline pkgs/applications/video/shaka-packager/default.nix:22]: Small improvement suggestion:

```suggestion
  };
  source = sources."${stdenv.hostPlatform.system}"
```

File: pkgs/applications/video/shaka-packager/default.nix
Comment (doronbehar) [inline pkgs/applications/video/shaka-packager/default.nix:30]:
```suggestion
    url = "https://github.com/shaka-project/shaka-packager/releases/download/v${finalAttrs.version}/${source.filename}";
```

**Review by doronbehar — CHANGES_REQUESTED** [review]
This is almost perfect! Sorry for getting so many iterations. I promise I'll merge this when after this is done!

**Comment by simonrw** [comment]
> This is almost perfect! Sorry for getting so many iterations. I promise I'll merge this when after this is done!

Thanks, better to get it right!

**Comment by doronbehar** [comment]
Result of `nixpkgs-review pr 231366` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>shaka-packager</li>
  </ul>
</details>

**Comment by doronbehar** [comment]
Tested `--help`, and seems functional.

**Review by doronbehar — APPROVED** [review]

**Comment by doronbehar** [comment]
@drupol : although your review requests have been addressed and the conversations after your review comments are marked as resolved, I can't merge this until you will approve. @simonrw : if @drupol won't respond in a few days we'll open a new PR and merge it quickly, promise. Sorry for this taking so long.

**Review by abathur — APPROVED** [review]
I can attest to general functionality working fine here. Using it in some WIP already.

**Review by drupol — APPROVED** [review]
Thanks for implementing all the changes!

**Comment by drupol** [comment]
Thanks again for your contribution, welcome to the Nix family !
Keep PRs coming :)

**Pull request merged**
